### ISSUE TITLE
Consider an option to ask for RPC target VM in dom0

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#910]
**Reported by marmarek on 28 Oct 2014 13:49 UTC**
This would allow for example copy some files out of VM, without telling the VM name of target VM.

In case of filecopy, it would not change user experience - user will still get a single prompt for VM name. In fact it will even simplify some things:
- can get rid of ugly VM name prompt on Windows
- can use the same script for nautilus and dolphin (no need to use zenity/kdialog to match the interface there)
- finally - can use the same prompt for VM name and service confirmation (when policy says "ask")

Technically, when source VM sends empty target VM field in service request, qrexec-policy would handle this case and show the prompt, instead of failing the request.

Migrated-From: https://wiki.qubes-os.org/ticket/910

**Comment by rootkovska** [comment]
See: https://groups.google.com/forum/#!msg/qubes-devel/xLZU0R5ijCg/_Lk1tXp9BwAJ

**Comment by marmarek** [comment]
Moving discussion from #2436:

I think the target VM argument should be preserved. There are multiple cases when you want to simply allow action for specific target VM without bothering user with confirmation. For example in 'devel' VM you may want to copy the outcome to some 'testing' VM where you run various tests. You don't want to select 'testing' VM hundreds times a day (you have a script to call qvm-copy-to-vm, or simply call it from shell history), but on the other hand you still may want to occasionally copy some files to a different VM.

This means I wouldn't touch the cases when policy explicitly says 'allow'. If you have policy `some-vm $anyvm allow`, it means that 'some-vm' VM will be able to learn list of VMs. If you don't want that, do not add such line to the policy.

The case of 'deny' action for existing vs non-existing target VM is easy - error message (if any) should not be blocking.

I should note that policy right now allow also to override target VM choice. You may specify something like:

    some-vm $anyvm allow,target=another-vm

Currently it means "whatever existing target VM name (if any) was provided, redirect service to `another-vm`". Should be: "whaveter target VM name is, redirect service to `another-vm`". Of course unless earlier rule says otherwise.

The thing to reconsider is what about 'ask' action. There are few cases:
 - calling VM provided target VM and it exists (and policy says 'ask')
 - calling VM provided target VM and it does not exist (but default action says 'ask')
 - calling VM did not provided target VM at all (and default action says 'ask')

Currently each of those cases are handled differently:
 - if target VM is provided and it exists you get the confirmation prompt, if it does not exists, you get (blocking) error message
 - if target VM is not provided, it is this ticket - user should be prompted to select one

I think non-existing target VM should be treated as no target VM provided at all (still - if the default action is 'ask').

So, we have two things to design:
 - what VMs should be listed in the prompt - initial idea was: all whose policy action would be "allow" or "ask".
 - (related to above) how to differentiate - if desired at all - between target VM selected in service call and target VM selected by the user in policy prompt - this IMO applies only to 'ask' action - should the policy prompt for specific target differ from prompt without target VM specified? Or maybe both should allow to select target VM, even if caller provided one?

And some UX issue: what should be the default target in that dialog (if any). If called provided one - I think it should be used), but otherwise - anything at all?
/cc @jpouellet

**Comment by jpouellet** [comment]
A first thought is to have the list of VMs always displayed, and the source-specified intended-target be the initial selection of that dropdown (or in case of text box: have it pre-populated with name of intended target). I don't know how I feel about that though, since I feel like it leads to building a user habit of simply pressing enter whenever source-suggested target exists, and what happens when one day the source is malicious and suggests a different target? User habit is still to press enter! Would they even notice the difference? Likely not...

Therefore, I think handling both "ask" cases (with and without suggested target) as if there were never a suggested target would be a good thing.

Case we need to consider is an `... ... ask,target=foo` rule. Is this even meaningful anymore? If so, which target should win (user specified vs. `target=`)?

**Comment by marmarek** [comment]
> and what happens when one day the source is malicious and suggests a different target? User habit is still to press enter! Would they even notice the difference? Likely not...

The point is to have only subset of VMs there - so it should be impossible to execute "dangerous" operation. Of course for that, someone would need to select which VMs should be there (set policy to 'deny' for others - or even 'deny' by default and 'ask' for a few). And the case you've described is exactly what we currently have - only a confirmation of what calling VM requested. I'm not saying it's the best possible, but not bad either. I think much bigger problem with it is focus stealing (#1166) - it's simple enough to guess when user will press enter and execute some qrexec operation just before...

> Another case we need to consider is an `... ... ask,target=foo` rule. Is this even meaningful anymore? If so, which target should win (user specified vs. target=)?

That's a good question. I'd say in that case, `target=` means what should be suggested, but the final decision is up to the user. Maybe this is the way to go: have target VM prefilled only when defined as such in policy, otherwise empty by default (even when specified by source VM)?

**Comment by jpouellet** [comment]
> I think much bigger problem with it is focus stealing

I agree, but my point was that even with perfect focus stealing prevention, pre-populating target in UI according to suggestion from (un-trusted!) rpc source seems to me a bad idea. It conditions users to accept the default, where the default is attacker-controlled.

> That's a good question. I'd say in that case, target= means what should be suggested, but the final decision is up to the user.

Sounds reasonable to me.

> Maybe this is the way to go: have target VM prefilled only when defined as such [target=] in policy, otherwise empty by default (even when specified by source VM)?

Yes. This sounds to me like the best option.

**Comment by boring-stuff** [comment]
Hi, I'd like to start on this.
Should I begin with implementing the Qt dialog first, the best place for this looks like [guihelpers.py](https://github.com/QubesOS/qubes-core-admin/blob/master/core/guihelpers.py)? I'd then integrate it in qrexec-policy and allow an empty target. 

Also, should I follow any particular Qt guidelines at all?

**Comment by marmarek** [comment]
Actually, GTK would be better, as new Qubes Manager would be in GTK, and also it will be more consistent with now default Xfce environment. I think it's better to create separate python file this GUI dialog.
Anyway the overall plan looks ok :)

**Comment by boring-stuff** [comment]
Noted. 
I've prototyped the window in glade and am hooking onto the elements via IDs, so that designers can later on change the .glade file without touching the code.
I'd also shoot for a re-usable python class that leverages a gtk combobox listing a subset of the domains, including colored icons. 

Question: how's the test strategy here? Are they all integration tests, or are there unit tests as well? How can I add testcases, and run them in a local environment? Are they all in [tests](https://github.com/QubesOS/qubes-core-admin/tree/master/tests), or somewhere else?

**Comment by andrewdavidwong** [comment]
Thanks for taking this on, @boring-stuff! I've added this to the [community-developed feature tracker](https://www.qubes-os.org/qubes-issues/#ask-for-rpc-target-in-dom0).

**Comment by marmarek** [comment]
> Question: how's the test strategy here? Are they all integration tests, or are there unit tests as well? How can I add testcases, and run them in a local environment? Are they all in tests, or somewhere else?

Yes, most tests are there. In Qubes 3.x we have only integration tests, but for Qubes 4.x (for which this feature is targeted), we also have unit tests. Take a look at core3-devel branch, especially here:
https://github.com/QubesOS/qubes-core-admin/tree/core3-devel/qubes/tests
There is also an API to have tests in other repositories ([here](https://github.com/QubesOS/qubes-core-admin/blob/core3-devel/qubes/tests/extra.py) - basically expose an endpoint for `qubes.tests.extra.for_template`), but I think it would be an overkill for this case.

I think at minimum we should have few tests calling `qrexec-policy` (directly) with various arguments:
 - valid source and destination domains, policy action `deny`, `ask`, `allow`
 - policy overriding target domain
 - various invalid values (service name, destination domain, invalid domain entered by the user etc)

Some existing integration tests (calling this through the whole stack - so from one VM to another):
- https://github.com/QubesOS/qubes-core-admin/blob/core3-devel/qubes/tests/int/vm_qrexec_gui.py#L322-L367
- https://github.com/QubesOS/qubes-core-admin/blob/core3-devel/qubes/tests/int/vm_qrexec_gui.py#L537-L644 

Also, IMO qrexec-policy tool should really belong to qubes-core-admin, not qubes-core-admin-linux repository. I'm going to move it there, but for your convenience you can continue to work on it in the old place.
Also, it's not so easy to have working working Qubes 4.0 installation to test stuff there ;) so you can work on this in Qubes 3.2 - while there are some API changes, should be minimal in this code.

**Comment by boring-stuff** [comment]
I understand, thanks for the hints! Is there a way of getting into a simulated environment without using an [extra physical machine](https://www.qubes-os.org/doc/test-bench/) for this?

As an example, I'd like to use the Gtk equivalent of getting a VM's icon (See [this](https://stackoverflow.com/questions/13353516/get-icon-name-for-a-stock-id) as an example), but when I run it in a VM it fails, since no such icons are installed in the templates. The icons are not even in the same repository, so there's no way to reference them directly. 
Another example is trying to import qubes.qubes which fails, since the main python files are under "core" until they get installed at the proper location.

Am I missing some way of setting up a testing environment? This is no blocker btw as I do use stubs for as long as I need to, but it would help...

**Comment by marmarek** [comment]
> I understand, thanks for the hints! Is there a way of getting into a simulated environment without using an extra physical machine for this?

Generally it's tricky... I haven't tried this for a long time. In theory it's easier for Qubes 4.0 (some unit tests can be run directly from git, so it works). See `run-tests` script in core3-devel branch.
Some approximation of separate physical machine may be a system started from separate partition or USB disk. Not very convenient, but for some development stage should be ok.

> As an example, I'd like to use the Gtk equivalent of getting a VM's icon (See this as an example), but when I run it in a VM it fails, since no such icons are installed in the templates. The icons are not even in the same repository, so there's no way to reference them directly.

You can add appropriate repository, then download artwork and other packages, then install them with `rpm -i --nodeps`. It's major hack, but for this particular case should work.

> Another example is trying to import qubes.qubes which fails, since the main python files are under "core" until they get installed at the proper location.

Similar here. You can do the same with locally built package. But `import qubes.qubes`, or try to loading `qubes.xml` will probably try to connect to libvirt, enumerate VMs etc, which will fail in VM. You can try setting `dry_run` at the beginning of qubes.py, but this mode isn't maintained for a long time...

**Comment by jpouellet** [comment]
@boring-stuff Thanks for tackling this. One less item on my personal to-do list :)

> Is there a way of getting into a simulated environment without using an extra physical machine for this?

So... partially. I wanted this too for various reasons. I had some success with running Qubes inside Qubes:
- In the outer Qubes, create an HVM vm, and install Qubes on it (advanced tab lets you attach an installer CD ISO).
- You can run only PV VMs inside that inner Qubes instance.

I did not get networking working (because sys-net requires working PCI passthrough of the emulated network device, and AFAIK qemu can't). It was annoying, but still enough for testing things requiring proper inter-VM communication.

**Comment by boring-stuff** [comment]
Thanks for the support and answers! I might try the HVM because from now on I'll be mostly blindly coding without it, but it might take some time.

First, a status report: the [gtkhelpers.py](https://github.com/boring-stuff/qubes-core-admin/blob/master/core/gtkhelpers.py) utility (tests [here](https://github.com/boring-stuff/qubes-core-admin/blob/master/tests/gtkhelpers.py)) now can show a Gtk confirmation window, depending on the parameters `source` and `target` (which is `None` by default).

The `core/gtkhelpers.py` along with [the `gtk` folder](https://github.com/boring-stuff/qubes-core-admin/tree/master/core/gtk) can be copied in the qubes module directly, and plug into it.

Usage and tests: 
```
from qubes import gtkhelpers

gtkhelpers.TransferWindow.confirm_transfer("sys-net")
gtkhelpers.TransferWindow.confirm_transfer("sys-net", "sys-firewall")
gtkhelpers.TransferWindow.confirm_transfer("sys-net", "non-existing")
gtkhelpers.TransferWindow.confirm_transfer("sys-net", "sys-net")
```

The user must/can then select (and eventually override, if `target != None`) a destination domain and confirm with `ok`. The function then returns e.g. `{'target_name': 'my-vm-name', 'target_id': 5}` with a confirmation, and False if cancel is pressed, or the window is closed.

**Note:** the unit tests are quite extensive, but they're not yet integrated in the build process. They don't perform any edit to the configuration whatsoever, but instead mostly use mock classes exactly to avoid that. Also the import should be fixed, depending on the environment in which they're run.

**Comment by boring-stuff** [comment]
Wait, this should actually affect all of the RPC actions, not just the filecopy, right? Then I'll first change the template and names to be a generic one, before starting with further integration...

**Comment by marmarek** [comment]
> Wait, this should actually affect all of the RPC actions, not just the filecopy, right? 

Yes...

**Comment by boring-stuff** [comment]
Updated my previous comment, names and semantic fixed.

**Comment by marmarek** [comment]
Generally looks good!
Minor comments:
 - can we have more structured message, like source domain and service name in table-like view, for example:
```
Source qube: <b>%s</b>\n
Action: <b>%s</b>\n
```
IMO that would make easier to quickly read the important parts.

 - target VM name selection should be search-able/filterable - to allow entering VM name by keyboard (or enter few characters, then select from list). You know, some of us have a lot of qubes ;)

**Comment by boring-stuff** [comment]
@marmarek could you elaborate on what you're looking for as a table view?

I'll be updating the [previous status comment](https://github.com/QubesOS/qubes-issues/issues/910#issuecomment-273154431) with ToDos and status as the tasks get done...

**Comment by marmarek** [comment]
> @marmarek could you elaborate on what you're looking for as a table view?

Something to ease quickly finding important information: what qube, what action, where. Currently one need to extract this from a text - bold helps somehow, but IMO it would be better to have this information below the question text in a structured form.

**Comment by boring-stuff** [comment]
Selection now searchable, confirm with the "enter" key is also supported.

To suppress the automatic text column that gets added if `combo.has_entry()` I clear the packed column before adding the new ones. As a result, however, I get this Gtk warning:

```
(gtkhelpers.py:28407): Gtk-CRITICAL **: gtk_cell_layout_set_attributes: assertion 'GTK_IS_CELL_RENDERER (cell)' failed
```

**Comment by marmarek** [comment]
Is is possible to reuse that text column instead?

Also a minor thing: ESC should cancel the dialog.

**Comment by boring-stuff** [comment]
I wanted to suppress the standard behavior, which has the text column on the left and the domain icon on the far right of the popup (it looked awkward), so of course we can default to that if wanted.
I'll try to re-use the column and move it on the right of the icon column, yes, the current implementation is more of a ugly hack than anything.

Noted on the ESC key... which brings me to the next topic.

@jpouellet @marmarek: the current implementation, when executed in dom0 with xfce, **does not** steal focus (this seems to be an effect of having set the window's `type_hint='notification'`). This behavior can be altered programmatically by raising the window to focus, but I think we should evaluate the up and downsides.

- With autofocus off, I can't alt-tab to the question, I have to click on it or click on the taskbar entry ([`set_skip_pager_hint(False)`](https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Window.html#Gtk.Window.set_skip_pager_hint) has no effect). Other windows can also get on top of it and hide it from view.
- If I use [`set_keep_above(True)`](https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Window.html#Gtk.Window.set_keep_above), the window stays on top of every other window, but also steals focus
- Any other valid `type_hint` steals focus, and many of them have the window shown in the alt-tab pager

What I think would be ideal, but I don't know how to achieve, is:
1. the dialog doesn't steal focus
2. alt-tabbing to it must be still possible

**Comment by boring-stuff** [comment]
Current status after the [last commit](https://github.com/boring-stuff/qubes-core-admin/commit/be806e9f7cdd44aea349d8710685da33b924a35d) can be seen [here](http://en.zimagez.com/zimage/rpc1.php).

**Comment by marmarek** [comment]
What package provides `gnome-foot` icon?

**Comment by marmarek** [comment]
In the meantime - new layout is exactly what I had in mind! Also, search, list with icon etc looks good.

One minor thing about search: ease selecting VM with given prefix - for example if I have VMs `work-web`, `work-mail`, `work-vault`, I'd like to enter `work-` (or a prefix of it), then select which of them I want to use (arrow down/up). But the current version isn't bad either - I can enter a prefix of `work-`, then right arrow, and enter prefix of remaining part (`w`, `m`, or `v`). Just can't easily list them (without scrolling that list). Yes, I have over 100 VMs... And I know people hitting hard limit of 256...

As of focus stealing, maybe some other approach, based on security-sensitive prompts in Firefox - steal the focus, but activate "OK" button not earlier than X sec after getting focus? X can be 1s or even 0.5s - something to give a chance to not click/press enter if the prompt appears as a surprise.
If not going that way, alt-tabbing to it is very important - as currently it's very easy to get the window under something else without an easy way to bring it back, or even know that something is waiting for confirmation there.

**Comment by jpouellet** [comment]
> As of focus stealing, maybe some other approach, based on security-sensitive prompts in Firefox - steal the focus, but activate "OK" button not earlier than X sec after getting focus? X can be 1s or even 0.5s - something to give a chance to not click/press enter if the prompt appears as a surprise.

I like this approach as well.

Additionally, I think it would be good if either the window were somehow pinned to the front, or if the OK button would be inactive within some time period from the window becoming active, not just being created, in order to mitigate the following:
- immediately before opening expected window in untrusted VM, initiate RPC dialog (in rapid succession such that user does not notice RPC dialog open underneath the expected window)
- invite user interaction with top window while RPC dialog "Ok" button disable times out
- position attractive UI element on top window directly above "Ok"
- when mouse is about to be clicked, close window
- user accidentally clicks "Ok"

and similarly for keyboard-driven "Ok" button activation.

**Comment by jpouellet** [comment]
@boring-stuff I have not found time to review your code yet, but IMO the interface in your screenshot clearly conveys the right information. :+1:

**Comment by boring-stuff** [comment]
@marmarek gnome-foot comes from gnome-icon theme. We could just raise a `ValueError` if an icon is not found in the productive code, whilst providing a dummy icon only for the unit tests, it's your call.

As for the autocomplete feature, I'm still targeting Gtk version 3.0, since that's the only one available in dom0:
```
$ for i in 3.0 3.2 3.4 3.6 3.8 3.10 3.12 3.14; do python -c 'import gi; gi.require_version("Gtk","'$i'");' &>/dev/null && echo $i supported; done;
3.0 supported
```
See, my code for autocomplete right now is already a bit of a hack, directly instructing the Gtk.ComboBox's child Gtk.Entry to do autocomplete, whereas starting from Gtk3.2+ more autocomplete features can be enabled centrally at the ComboBox level, which I expect would better handle the arrows problem as well. So either:
- we update libgtk to a newer version and target that instead (would have to be tested whether that would add the functionality)
- we hack a bit more into it, which might not be ideal (I suppose we might have to override more stuff from the combobox)

As for the focus problem, I see we have a general understanding:
- Window should steal focus: `type_hint!='notification'` should do it already
- Window should be pinned to the front: `set_keep_above(True)` as soon as the window gets visible
- Window should appear in the pager and taskbar
- "OK" button should not be active even with a correct target for 1s after windows is focused. I'll start by having it inactive after shown, will probably need to hook onto a "focus" signal to have it better defined though.

(btw. note that, as expected, we are at the WM's mercy, windows can only ask or signal for the first three features, but can be overridden)

@jpouellet I'll start splitting the code in the right modules, then, since right now everything's in one "gtkhelpers" python file. there are a couple of helpers (`VMListModeler.apply_model()` can be used on any Combobox or Treeview to list qubes, and likewise `VMListModeler.apply_icon()` sets a qube's icon on an Entry. GtkIconGetter, instead, is a helper to get and cache stock icons of a certain size), but the RPC code should get separated from that...

**Comment by marmarek** [comment]
> OK button would be inactive within some time period from the window becoming active, not just being created

Yes, that's why I've written "activate "OK" button not earlier than X sec after getting focus". And should be deactivated when window looses the focus.

As for GTK version, that looks like just another reason for upgrading dom0 in Qubes 4.0 (#2574).
Currently in dom0 `gtk3` package version is 3.18. In Fedora 25 it's 3.22. BTW it looks like the correct [method of checking GTK version is this](http://stackoverflow.com/a/11099204):
```
Gtk.get_major_version()
Gtk.get_minor_version()
Gtk.get_micro_version()
```

**Comment by boring-stuff** [comment]
Just to be sure, those versions (3.18 and 3.22) are not the Gtk versions that can be imported in python with `gi.require_version`, for that I guess other packages must be updated as well, I don't know the available version in Fedora 25, though.

**Comment by marmarek** [comment]
For me the version you specify in `gi.require_version` is not exactly a GTK version - maybe only major version or so. When you do that, the imported `Gtk` modules still report 3.18 (or 3.22 in fc25) using the method I've given above.

**Comment by boring-stuff** [comment]
I see. Confusing, but it makes sense, I guess.

I've solved the selection of VMs with the [latest commit](https://github.com/boring-stuff/qubes-core-admin/commit/773145cf7acd66a32d0666ce36b7917f92444c74) ([here](http://en.zimagez.com/zimage/rpc20.php) a screenshot)

**Comment by marmarek** [comment]
Should the delayed activation of "OK" button already work? I've tried (in a VM, not dom0) and it doesn't...

**Comment by marmarek** [comment]
Besides this, the GUI part looks awesome! :)
I've briefly looked at the code, and have few comments:
 - remove trailing spaces
 - don't use bare `except` - specify exception class

In Qubes 4.0 we have pylint to check for such things (and much more), but the code for Qubes 3.x is too ugly for that :/

**Comment by boring-stuff** [comment]
@marmarek yes, with commit [58ed312d - "FocusStealingHelper protects against focus stealing"](https://github.com/boring-stuff/qubes-core-admin/commit/58ed312d3d34ca558d465cd9d45087f9b21e69b3), the following can be seen in a VM after the window gets focused:
- OK button remains disabled for 1s, even if valid target is specified (i.e. with a parameter)
- Cancel button is enabled but does not work for 1 second
- ESC does not cancel the dialog for 1 second
- As soon as the window de-focuses, the status is reset (button and actions disabled)

Windows in VMs do not receive any focus/unfocus events from the WM after being shown, so this behavior can only be tested when the window is started. Have you tested with an earlier version? Starting `tests/rpcconfirmation.py -w` should show the behavior above.

I'll check pylint then and fix the code, using a static analysis tool seems like a good idea overall.
For the next steps, I'm taking apart the `qrexec-policy` code (from the master branch of `qubes-core-admin-linux`) and making it more object oriented (i.e. one class only to find/read a policy file) in order to test the different components separately to then let integration tests check the complete workflow. I'd also start with an helper module in `qubes-core-admin`, since I understand you plan to eventually move `qrexec-policy` there completely. Does that sound right?

**Comment by marmarek** [comment]
Ah, indeed it works now - had earlier version before. It's interesting why disabling doesn't work in the VM - at least in theory window do receive such events:
```
[user@testvm ~]$ xev -id 0xe00008


FocusIn event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyNonlinear

KeymapNotify event, serial 18, synthetic NO, window 0x0,
    keys:  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

FocusOut event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyInferior

FocusOut event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyNonlinearVirtual

ConfigureNotify event, serial 18, synthetic NO, window 0xe00008,
    event 0xe00008, window 0xe00008, (2043,414), width 400, height 242,
    border_width 0, above 0x200002, override NO

FocusIn event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyNonlinear

KeymapNotify event, serial 18, synthetic NO, window 0x0,
    keys:  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

FocusOut event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyInferior

EnterNotify event, serial 18, synthetic NO, window 0xe00008,
    root 0x27e, subw 0x0, time 47210167, (146,0), root:(2189,414),
    mode NotifyNormal, detail NotifyNonlinear, same_screen YES,
    focus NO, state 0

KeymapNotify event, serial 18, synthetic NO, window 0x0,
    keys:  126 0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

LeaveNotify event, serial 18, synthetic NO, window 0xe00008,
    root 0x27e, subw 0x0, time 47213603, (150,3), root:(2193,417),
    mode NotifyNormal, detail NotifyNonlinear, same_screen YES,
    focus NO, state 0

FocusOut event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyNonlinearVirtual

ConfigureNotify event, serial 18, synthetic NO, window 0xe00008,
    event 0xe00008, window 0xe00008, (2043,414), width 400, height 242,
    border_width 0, above 0x200002, override NO

EnterNotify event, serial 18, synthetic NO, window 0xe00008,
    root 0x27e, subw 0x0, time 47215786, (141,7), root:(2184,421),
    mode NotifyNormal, detail NotifyNonlinear, same_screen YES,
    focus NO, state 0

KeymapNotify event, serial 18, synthetic NO, window 0x0,
    keys:  126 0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

FocusIn event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyNonlinear

KeymapNotify event, serial 18, synthetic NO, window 0x0,
    keys:  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

FocusOut event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyInferior

LeaveNotify event, serial 18, synthetic NO, window 0xe00008,
    root 0x27e, subw 0x0, time 47216541, (3,75), root:(2046,489),
    mode NotifyNormal, detail NotifyNonlinear, same_screen YES,
    focus NO, state 0

FocusOut event, serial 18, synthetic NO, window 0xe00008,
    mode NotifyNormal, detail NotifyNonlinearVirtual
```

Not sure if the "cancel" behavior is right - if it's disabled for 1s, the the button should also be inactive (otherwise it's confusing). But IMO there is no need to disable cancel action in the first place. @andrewdavidwong, @jpouellet  any opinion?

As for `qrexec-policy` - current structure also allow such usage, as policy loading, matching etc are all separate (quite independent) functions. Keep in mind that this whole feature is for Qubes 4.0, and `qrexec-policy` is already slightly different there - see `core3-devel` branch on my github account. As I'll need to plug your work into new codebase anyway, either don't restructure `qrexec-policy`, or use the new version (I prefer the former).

**Comment by jpouellet** [comment]
> - Cancel button is enabled but does not work for 1 second
> - ESC does not cancel the dialog for 1 second

I am in favor of mitigating accidental *approval* of RPC, but what is the purpose of disallowing quick canceling?

This sounds annoying IMO.

**Comment by jpouellet** [comment]
Looks like @marmarek beat me to it.

> But IMO there is no need to disable cancel action in the first place. @andrewdavidwong, @jpouellet any opinion?

I agree. Prefer not disabling cancel.

**Comment by jpouellet** [comment]
@marmarek perhaps not working in a VM is related to https://github.com/QubesOS/qubes-issues/issues/1455?

**Comment by marmarek** [comment]
> @marmarek perhaps not working in a VM is related to #1455?

Related to `set_keep_above(True)`? Simple experiment (by commenting out this line) does not confirm this theory. But indeed in real dom0 disabling "OK" does work.

**Comment by andrewdavidwong** [comment]
> But IMO there is no need to disable cancel action in the first place.

I can understand the UX rationale for the delay (e.g., accidental key presses), but I don't have a strong opinion about it. If not clearly indicated, it can seem like a bug. In the end, though, a 1 second delay is not a big deal either way.

**Comment by rootkovska** [comment]
I think the 1sec delay is a good compromise to achieve reasonable anti-focus-stealing attacks. Especially that all WM I tried didn't implement this anti-stealing in a satisfactory way.

They GUI sample looks really nice, BTW :)

**Comment by rootkovska** [comment]
Also, I think that, in the interest of clarity and simplicity (as well as user-non-confusing) we should ignore any dstvm arguments passed by the srcvm.

> I think the target VM argument should be preserved. There are multiple cases when you want to simply allow action for specific target VM without bothering user with confirmation. For example in 'devel' VM you may want to copy the outcome to some 'testing' VM where you run various tests. You don't want to select 'testing' VM hundreds times a day (you have a script to call qvm-copy-to-vm, or simply call it from shell history)

So, in that case one should provide a specific policy:

`devel testing my.TestService1 allow`

> but on the other hand you still may want to occasionally copy some files to a different VM.

This case could be then resolved by using specific services for automated tests, as I exemplified above.

**Comment by marmarek** [comment]
> Also, I think that, in the interest of clarity and simplicity (as well as user-non-confusing) we should ignore any dstvm arguments passed by the srcvm.

For `ask` action - yes this is what we're going to do - ask for the target VM in that confirmation dialog, instead of source VM (actions like file copy etc). [Even if source VM provided some](https://github.com/QubesOS/qubes-issues/issues/910#issuecomment-260817960) (which resolved to `ask` action). In Qubes 4.0 we allow empty target domain in qrexec service call - exactly for this reason - as it doesn't make sense to send something that would be ignored anyway (for services where the major use case is action `ask`).

But not for allow/deny - if your policy specify allow for some destination domains, it should be allowed to do so. Consider `qubes.Gpg`, for example policy like this one should be perfectly valid policy:
```
project-x keys-project-x allow
project-x keys-project-x-testing allow
test-environment keys-project-x-testing allow
```
I think it would be really bad idea to force the user to duplicate a service for such cases.

> So, in that case one should provide a specific policy:
> 
> devel testing my.TestService1 allow

Can you elaborate this syntax? This isn't how our policy looks like.

**Comment by rootkovska** [comment]
> Can you elaborate this syntax? This isn't how our policy looks like.

Fixed in the previous msg.

**Comment by rootkovska** [comment]
Do we have more examples of services like qubes.Gpg where we would like to have allow-policies (instead of 'ask') for more than one destVM for the same srcVM?

**Comment by marmarek** [comment]
Ok, so I think we agree on this:
 - keep dest VM _at protocol level_
 - ignore it when asking the user for confirmation (start with no default dest VM, unless _policy_ specified one[*])
 - respect it when policy explicitly specify allow/deny

[*] Such policy would look like:
```
vm1 $anyvm ask,target=default-target
```

> Do we have more examples of services like qubes.Gpg where we would like to have allow-policies (instead of 'ask') for more than one destVM for the same srcVM?

Yes, for example this (still my.TestService1):
```
devel test-debian-8 allow
devel test-debian-9 allow
devel test-windows allow
```
Or this - `qubes.OpenURL`:
```
$anyvm $dispvm:dispvm-firefox allow
$anyvm $dispvm:dispvm-chrome allow
```
Or even for `qubes.Filecopy`:
```
# chosen by a different key in mutt for example
personal-mail personal-files-current allow
personal-mail personal-archive allow
```
Or this - for qubes.ClipboardPaste:
```
# note that `ask` means really `allow` here, as Ctrl-Shift-V is treated as implicit confirmation
$tag:work $tag:personal deny
$tag:work $tag:work ask
```
Do you need more examples? I think I can produce dozen more.

**Comment by marmarek** [comment]
And BTW, if you want, you can also specify target VM at policy level, ignoring what calling VM have sent, also for `allow` action:
```
srcvm $anyvm allow,target=dstvm
```
And also have "just confirmation" with pre-filled target VM:
```
srcvm $anyvm ask,target=dstvm
```
Both already works in 3.2.

**Comment by rootkovska** [comment]
Sure, I do understand the appeal of what you're proposing. Appeal for both advanced user as well as advanced integration. But from the security perspective I also find it very attractive each time we have a chance to get rid of the necessity to parse an understand input.

I'm sure you would say that: 1) we already parse that in Qubes 3.x, and 2) this is "just" one argument and we know how to parse it carefully and correctly. But: 1) we can always improve over what we have and it seems like this is a good opportunity to do so now. And, 2) maybe it is true that it is simple to parse this destVM argument now, but it is thinkable that in the future the policy might become even more complicated. Perhaps some more special cases will be added. Or perhaps somebody will find a surprising attack: e.g. that if a OOM happens in Dom0 precisely when the code is consulting the policy to see if destVM should be ignored or not, then it will be decided for it to be actually not ignored. I'm sure we can think of more _hypothetical_ cases here. In any case, it just strikes me as confusing that the destVM should be ignored in some cases, while used in some other cases. This surely smells like troubles...

So, perhaps it's worth considering these exemplary cases that you listed and to see if they couldn't be easily worked around, without needing to conditionally use the destVM argument? Let's see:

1. The automation of some development workflows:
> Yes, for example this (still my.TestService1):
> devel test-debian-8 allow
> devel test-debian-9 allow
> devel test-windows allow

In this case, I think it just easy to create dedicated services (my.TestService{Debian8,Debian9,Windows}) and install them in corresponding templates.

But i'm thinking this could be made even simpler if we made use of our "policy argument". I.e. we could have only one service (in the destVM), namely: myTestService, but policy would be defined using the the additional argument, e.g.. in the `/etc/qubes-rpm/policy/my.TestService+Debian8` we will have:
`devel test-debian-8 allow`

2. > Or this - qubes.OpenURL:
> $anyvm $dispvm:dispvm-firefox allow
> $anyvm $dispvm:dispvm-chrome allow

How about: `qubes.OpenURL+Chrome` and `qubes.OpenURL+Firefox`?

And the same for qubes.Filecopy and qubes.Gpg examples you gave.

3. > Or this - for qubes.ClipboardPaste:
> note that `ask` means really `allow` here, as Ctrl-Shift-V is treated as implicit confirmation
> $tag:work $tag:personal deny
> $tag:work $tag:work ask

Admittedly, I will have to think more about this special case.

BTW, one can say it is cheating, because we still smuggle an untrusted argument into policy-parsing process. However: 1) this argument is treated much more uniformly than the destVM would be -- there are no special cases when to ignore it and when to not. It is always parsed the same way. 2) it's mishandling is potentially much less dangerous, because the destVM is still one that the user allowed in the policy (but perhaps for other version of the service). Finally 3) we don't have any options to get rid of this argument ATM anyway, so it's going to be there anyway :)

**Comment by rootkovska** [comment]
As for the `qubes.ClipboardPaste` case: because the destVM argument is always trusted (it's provided by the guid), and because we already have a special-case hard-coded call to qrexec-policy from within guid to handle this special case (by passing the `--assume-yes-for-ask`) I think this case should not present a problem. We can always pass the desired (and trusted) destVM to qrexec-policy from within guid.

**Comment by marmarek** [comment]
> In this case, I think it just easy to create dedicated services (my.TestService{Debian8,Debian9,Windows}) and install them in corresponding templates.

While this would work in this particular case, maintenance of it would be a nightmare - consider someone packaging such services - then you'd constantly receive support requests like "it doesn't work on Debian 10!". It would be very similar to maintenance of split-gpg... ("new version of enigmail require those 3 new options").

> But i'm thinking this could be made even simpler if we made use of our "policy argument". I.e. we could have only one service (in the destVM), namely: myTestService, but policy would be defined using the the additional argument, e.g.. in the /etc/qubes-rpm/policy/my.TestService+Debian8 we will have:
> devel test-debian-8 allow

And you just moved exactly the same string ("destVM") verified in exactly the same way (list of allowed values), to a slightly different place - but still handled in dom0. And actually abusing policy argument for this is worse than using destination domain: destination domain is enforced by the qrexec-policy tool to be a valid qrexec target (existing domain, `$dispvm`, or, in Qubes 4.0 - empty), whatever you write in the policy[*]; but it is possible to write a policy accepting any "policy argument" (which is then passed to a the target domain) - simply write a policy for `some.service`, not `some.service+argument`.

[*] You can't, by design, abuse target domain for example using abstract names redirected later to actual domains. That value is checked before even looking at the policy, as part of validation, just after checking for having only allowed characters, not being too long etc. This feature allow (in case of some policies) the caller to brute-force list of VMs. But the fix is easy (https://github.com/QubesOS/qubes-issues/issues/910#issuecomment-260759464) - treat invalid target VM name as empty one and let the policy decide what to do (see below). This would still preserve the validation, understood as "don't pass further value not being one of allowed values".

> How about: qubes.OpenURL+Chrome and qubes.OpenURL+Firefox?
> 
> And the same for qubes.Filecopy and qubes.Gpg examples you gave.

Same response as above: abusing policy argument here is worse than using target domain.
Using policy argument for a browser choice may make sense for totally different reason - in that case it could the same (Disp)VM.

If your main objection is target domain being ignored _sometimes_, we can take a different approach:
 - Do not ignore target domain - even in case of `ask` action use it as pre-filled value in confirmation dialog.
 - Adjust policy for services like `qubes.Filecopy` to disallow any specific target domain specified by the caller and only allow the empty one. Such policy would look like this:

        $anyvm $empty ask

    instead of:

        $anyvm $anyvm ask

A downside of this approach is that we can no longer use qrexec policy to limit what target choices are available in that confirmation window, at least not in non-confusing form. In case of initial idea, that was easy:

    work-devel work-archive ask
    work-devel work-mail ask
    work-devel $empty ask
    work-devel $anyvm deny

Then confirmation prompt (when is call initiated by `work-devel`) would have two available options: `work-archive` and `work-mail`. In case of using caller-specified target domain for initial value in the confirmation dialog *and assuming we do want to forbid it in some/majority cases*, it would no longer work. Because specifying `work-devel work-mail ask` would allow the prompt with `work-mail` as default value.
While I can think of some syntax that would workaround that problem, IMO it would be confusing.

Side note: `$empty` keyword means "caller didn't pass any target domain".  It isn't specified anywhere yet. Initially I was thinking about using `$default`, but now I think `$empty` would make more sense.

The current proposal (with ignoring target domain) is simple:
 - action `allow`: allow the call to the domain specified by the caller, potentially overriden by the `target=` argument in the action (same as in Qubes 3.2)
 - action `deny`: deny the call, including possible further `ask` actions (don't list this option as possible value)
 - action `ask`: ask the user for the target domain and confirmation, use `target=` argument (if present) for default value

**Comment by boring-stuff** [comment]
Adding my 5 cents to the security topic, do tell me if I'm not seeing something here.

First, if we go down the road of parsing the `destVM` parameter, I'm a fan of parsing either all or none as a default. I mean that if the `destVM` that we parse is mentioned in the qrexec policy, then fine, we proceed to apply the corresponding rule. If the `destVM` value equals a fixed constant such as `$none`, then we can ask for a target with the UI. Any other value should trigger a visible error message, and not proceed at all.

As for the bigger question of whether to keep the parameter at all:

### 1. Parsing the `destVM` parameter
Pros | Contra
-------------------------- | --------------------------
Easy to develop rules in the policy file for automating tasks | Interaction between VMs is defined in the VMs and is less visible in dom0 (renaming a VM would not register in the VM specifying the old target)
### 2. Removing the `destVM` parameter altogether
Pros | Contra
-------------------------- | --------------------------
Potential attack source removed, only the "+Param" parameter can be used to specify the destination VM | Apps in the VM calling services must support using a "+Param" in the service name, which becomes less manageable
Management of targets completely done in a secure environment dom0, a VM does not have to know of any other VM's name | 

To sum it up, option 1 leaves the mechanism of automated services quite easily manageable and scriptable, whereas option 2 closes a potential future exploit vector, but at the expense of users with automated services, who now have to manage them in the VMs and dom0 separately.

My next question would be: how many users out there have a need for automated qrexec services  (typical Qubes users, but which might include companies looking into `salt`y ways to configure the system), and how many services do we expect to run automatically?

**Comment by jpouellet** [comment]
> If your main objection is target domain being ignored sometimes, we can take a different approach:
> - Do not ignore target domain - even in case of ask action use it as pre-filled value in confirmation dialog.

I am still opposed to this for the same reason I pointed out in https://github.com/QubesOS/qubes-issues/issues/910#issuecomment-260815073 (builds dangerous user habit of just clicking ok, accepting untrusted destvm field populated by untrusted input)

**Comment by rootkovska** [comment]
I discussed a bit more with Marek in private (in the interest of not polluting this forum) and here's what we have concluded thus far. I will refer to my proposal presented above as "service-oriented" qrexec, and to what Marek has been proposing (and what is essentially only a slight variation from what we have had in Qubes all the time) as "classic" qrexec. First the summary:

1. In the classic qrexec the source VM explicitly requests which target (destination) VM should handle the service. This can be overridden by specifying `target=some_other_vm` in the policy (in Dom0). Additionally, and this is what is being proposed to change now, as a result moving the target-asking-dialog into Dom0, whenever the policy says 'ask', then the target VM provided by the source VM will always be ignored and the system will use whatever the user provided in the (Dom0-managed) dialog box.

2. In the "service-oriented" qrexec, I propose that source VMs never need to be aware of who is going to serve their qrexec request. They only would require a specific service and it will always be up to the policy (or the user clicking in Dom0) to select the target. Never for the source VM. So the `qrexec-client-vm` would no longer take the `target_vmname` argument.

So, instead of writing, as in the "classic qrexec":

`qubes.Gpg:`
```
work-email gpg-keys-work allow
work-email gpg-keys-special allow
```
... we would write in "service-oriented qrexec" as (for brevity I "collapsed" the notation):
1. `qubes.Gpg+work: work-email gpg-keys-work allow`
2. `qubes.Gpg+special: work-email gpg-keys-special allow`

Or, instead of:
`qubes.OpenURL:`
```
$anyvm $dispvm:dispvm-firefox allow
$anyvm $dispvm:dispvm-chrome allow
```
... we would state in service-oriented case:
1. `qubes.OpenURL: $anyvm $dispvm allow` (default when no specialized service called)
2. `qubes.OpenURL+Firefox: $anyvm $dispvm:dispvm-firefox allow` (use dispVM with Firefox installed)
3. `qubes.OpenURL+Chrome: $anyvm $dispvm:dispvm-chrome allow` (use dispVM with Chrome installed)

Now, let's review the pros and cons of both approaches:

1. We shall note that, since Qubes 3.2, it is _already_ possible to write policies using the "service-oriented" approach, specifying a dummy target VM, and (ab)using the `target=` modifier in the policy to redirect to the actual VM.

However, the (ab)use of `target=` is not elegant. Additionally, there is nothing that _enforces_ this kind of approach when creating a policy, so we can expect that what we ended up with would be a mixture of both approaches :/

Here, I think the service-oriented approach is a clear winner.

2. With the service-oriented approach we face a problem if we had more than one `allow` rules for a given (srcVM, service) tuple, e.g.:
`qubes.Mgtm`:
```
mgtmvm work1 allow
mgtmvm work2 allow
```
... or a single rule, but applicable to a group of VMs, rather than one, e.g. (tag support is coming in 4.0):
`qubes.Mgtm: mgtmvm $tag:work allow`

In such a case the policy evaluator in Dom0 doesn't know which target VM to use. In the first case we could -- following our current rule of "apply the first-matching policy" always direct qubes.Mgtm calls to `work1`. In the latter case, we would need to ask the user (dialog box). This might be inconvenient for some integration services, like the hypothetical "management service" used here for illustration.

3. In practice, however, the _actual_ management service that we're introducing in 4.0 is different (https://www.qubes-os.org/doc/mgmt1/). The service backend runs in Dom0, and the target VM that the source VM (the management VM) specifies, is treated as an _argument_ for the backend running in Dom0. This is implemented using the `target=dom0` modifier in the policy for all of the management calls.

Now, this "creative" use of the target VM argument can be seen as both inelegant and elegant :) 

* It is arguably inelegant and confusing, because, the mgmt service is actually running in Dom0, not in whatever VM is given in the target argument. This "override" of the interpretation of such an important argument makes for a dangerous precedent (even if it is rather clearly visible the override is taking place, thanks to the `target=dom0` specifier).

* It's arguably elegant, because it allows to neatly keep all the policy within one place, rather than spread between the qrexec policy (which would only say that some mgtm VM can perform management calls to Dom0) and the actual policy for the management backend (which would now specify which VMs should be manageable from each management VM(s) and what types of mgmt calls are allowed).

E.g. one can specify a policy directly for `mgmt.vm.property.Set` or even for `mgmt.vm.property.Set+netvm`, thus controlling which or whether a mgmt VM can change the netvm for some other VM(s). 

4. The "service-oriented" approach has one big, practical disadvantage: it breaks backwards compatibility for most existing qrexec services. While the fix would usually be very simple: just provide more specific service description (using `+something`), it nevertheless would need to be done.

**Comment by marmarek** [comment]
Yes, this is very accurate description of our discussion. I'd add two points:
1. As you mentioned - it is possible to implement "service-oriented" approach using "classic" approach, but not the other way around. Since writing a qrexec service require some security-focused design (and implementation!) anyway, it's reasonable to assume that such person is capable to choose the right model for a given service. And we can include some more guidelines in the [documentation](https://www.qubes-os.org/doc/qrexec3/).
2. Backward compatibility is important factor - we've managed so far to have multiple qrexec services compatible across multiple Qubes versions - for example split gpg (master branch) is compatible with all Qubes versions since 1.0. This is very helpful for maintaining those services. Breaking this compatibility means we'd have to maintain parallel version of all `qubes-app-*` components.

A minor note: I do not agree that `target=` parameter is inelegant syntax - it's a logical order of action line. For example `qubes.Gpg`:

    (1)work-email (2)$anyvm (3)allow,target=(4)gpg-keys-work

you read as: when `qubes.Gpg` called from `work-email`(1) with any destvm(2), allow(3) it and direct it to `gpg-keys-work`(4). Note that we have clearly ordered conditions first, actions later.
(I've used `$anyvm` since `$empty` isn't part of the specification yet)

But in "service-oriented" approach, with the current syntax:

    (1)work-email (2)gpg-keys-work (3)allow

you read as: when `qubes.Gpg` called from `work-email`(1), allow(3) it and direct to `gpg-keys-work`(2). Here conditions and actions are mixed.
As "service-oriented" approach is incompatible with the current one, we could probably also alter the syntax (swap columns "action" and "destvm") to fix this issue. So it's a minor argument against. But it would break compatibility even further, also making it confusing for service writers.

BTW I'm curious how many people have custom qrexec services ;) On one of my machines I have 15 of them.

**Comment by jpouellet** [comment]
> Now, this "creative" use of the target VM argument can be seen as both inelegant and elegant :)
> 
> - It's arguably elegant, because it allows to neatly keep all the policy within one place, rather than spread between the qrexec policy

target= may be in some ways inelegant, but it is easy to understand and allows one to naturally express intent, and IMO most importantly it provides a single centralized implementation of access control which is adequate for most (all?) expected use cases.

Continuing to "abuse" the feature in the current way means that there will continue to be only *one* piece of code responsible for determining inter-vm policy, rather than encouraging service implementations to roll their own and thereby increase auditing burden.

Having a centralized "*this vm* can invoke *this rpc* on *that vm*" mechanism eases both auditing of access control implementation(s), but also auditing of per-machine policies since there is only a single thing to parse. I can easily imagine a graphical tool which produces a graph of which domains can do what to other domains, and also the ability to statically assert that there is no chain of services which allows a particular vm A to in any way reach vm B. This would not be possible if each service wanting such ACL-like functionality is left to roll their own implementation.

**Comment by jpouellet** [comment]
I feel like I must somehow misunderstand the proposal or fail to see a major benefit or something.

My understanding is that the idea just to replace the following pattern:
```
source dummy_dest allow,target=actual_dest
```
with
```
source actual_dest allow
```
with the observation that in most cases dummy_dest is not inherent to the action we wish to perform, but rather is just an arbitrary VM with suitable programs installed (for example when we wish to view a PDF).

And in the case where the dest **is** somehow inherent to the action we wish to perform (for management, etc.) we expect to force the argument to be the name of the dest vm, force the service implementation to provide its own evaluation of whether or not source is allowed to operate on dest (which means the service must be in dom0)?

And do I understand correctly that @rootkovska's assertion is that that the benefit provided by a slightly simpler evaluation of policy in the usual case where dest is not inherent to the operation, outweighs the introduced issues of:
1. not being able to use the argument as an actual argument because it must be used to specify dest if we want to be able to express access control in the policy file
2. breaking backwards compat
3. cases such as `$anyvm:work $anyvm:work allow` requiring either N policy files with +workvm-1, +workvm-2, +workvm-3 (which need to be maintained appropriately), or requiring a component of the RPC implementation to be responsible for evaluating ACLs for tags by itself, and additionally be in dom0 to then invoke the actually desired operation in the evaluated target vm

in the case where dest is somehow inherent to the operation being performed?

This seems like a rather one-sided comparison to me, so I feel like I must be missing something. @rootkovska?

**Comment by jpouellet** [comment]
After enumerating my services [here](https://groups.google.com/forum/#!topic/qubes-users/60FoSFWkiNs) and thinking about them some more, I realize that all of them fall into one of the following categories:
- only have one destvm which could possibly be appropriate
- only in practice used for `$dispvm` targets, where having separate policies for `$dispvm:somekind` `$dispvm:otherkind` would be non-problematic
- makes more sense to specify the destvm via user selection in dom0 and have the calling context be unaware of it

So besides management, I think I now agree that the cases where the downsides of the service-oriented proposal actually appear are sufficiently rare.

**Comment by marmarek** [comment]
And some of mine services [here](https://groups.google.com/d/msgid/qubes-users/20170129020422.GU1285%40mail-itl). As you can see - most of them indeed have a single destination VM in practice, but there are few useful exceptions.

There is also one exception in qubes-builder related services: `qubesbuilder.CopyTemplateBack`. The service is used to send template built in DispVM (root.img) back into calling VM. Currently the policy looks like this:
```
$anyvm build1 allow
$anyvm build2 allow
...
```
This is mostly because the current policy does not allow to identify particular DispVM - the intention here is to "allow the service from DispVM back into the VM who created this DispVM" and currently it's achieved artificially using authorization token inside the connection. This one potentially could be improved when/if we'll have `$tag:dispvm-created-by-x`.

**Comment by rootkovska** [comment]
TBH, the "only" point that seems really problematic for me with this new "service oriented" qrexec, is the breaking of backward compatibility. But that's a very serious problem, and thus I think that for 4.0 we should stick to the "classic" qrexec model (but of course using the Dom0-based dialog for "ask" policies). 

Meanwhile, we should:
1. Attempt to (re)write services and policies according to the "service-oriented" model whenever possible,
2. Gather more experience with our new mgmt infrastructure, how it is being used *in practice*, both for various corporate management deployments (which we hope we will have a few ;) and things such as upcoming GUI domain (which we hope to introduce in 4.1).

Then, perhaps in Qubes 5.0 we should reconsider again to switch to the pure service-oriented model.

**Comment by boring-stuff** [comment]
Ok, the code seems pretty much ready for a review, pylint does not show any errors or warnings anymore.
The next steps would be (1) the integration in `qrexec-policy` and (2) specifying the list of VMs to be shown in the combo (if a restriction is needed). I would leave task 1 to you, only to implement 2 when you have a better idea of what and how to pass to the rpc confirmation window for filtering. Does that sound OK?

**Comment by jpouellet** [comment]
Would you like comments here? On the commits? In some PR?

**Comment by jpouellet** [comment]
Also, @boring-stuff is your key (0x4753F38CBC279FB5?) available someplace? I don't see it on the usual keyservers...

**Comment by boring-stuff** [comment]
I personally don't mind reading comments here, or do you have a better idea? If the code seems mature enough and you want to proceed integrating it in a branch somewhere, simply do so (eventually rebasing, if desired) and I'll submit patches when needed.

I've generated a key as soon as I created the github account and started coding (although one of the first commits is not signed, I hadn't set the automatic signature in my config). I've just published [the public key](http://qdigse2yzvuglcix.onion/pks/lookup?search=0x4753F38CBC279FB5&op=vindex).

**Comment by marmarek** [comment]
@jpouellet does it mean you want to help with reviewing this?

**Comment by jpouellet** [comment]
@marmarek I've made a first pass already, just blocking on available free time. Hopefully I'll get a chance tomorrow.

Will definitely want a final review when integrated w/ qrexec-policy. I plan to resolve https://github.com/QubesOS/qubes-issues/issues/2460 in the process.

**Comment by boring-stuff** [comment]
I'm looking at the receiving end of a file transfer (where #1793 would apply). Does it make sense:
1. that a destVM always known what origin VM started the transfer? (the [QREXEC_REMOTE_DOMAIN environment variable](https://github.com/QubesOS/qubes-core-agent-linux/blob/be0e8a250f5cb0698120a98ea6f9be34fc12d9cf/qubes-rpc/qfile-unpacker.c#L48) carries this information)
2. that it's not possible for the destVM to first select a destination folder before unpacking the file?
3. that no confirmation or success message is displayed in the destVM, with a link to the destination file?

I'm thinking that n.3 would help usability, but the obvious approach of opening the destination folder in a file manager would bring risks we should avoid (think of a file exploiting a thumbnailer). N.2 could be made optional by having a checkbox in the dom0 confirmation dialog an passing a "custom folder" parameter to the receiving agent. Note also that 2 and 3 require some sort of graphical interface, so they would need be not mandatory for every VM installation (so that minimal images are still possible).

**Comment by jpouellet** [comment]
> 01. that a destVM always known what origin VM started the transfer? (the QREXEC_REMOTE_DOMAIN environment variable carries this information)

Yes. That is [documented in the qrexec3 API](https://www.qubes-os.org/doc/qrexec3/), so I do not see it going away any time soon (because backwards compat).

> 02. that it's not possible for the destVM to first select a destination folder before unpacking the file?

sourcevm obviously must not. destvm may, but it's still dangerous. IMO the UX improvement isn't worth the danger of files ending up where they are not expected or where they have unwanted side-effects. I could be convinced otherwise, but it'd need to be a much stronger argument than "because we can". Only allowing files to land in ~/QubesIncoming/${vm}/... is a nice property from a security perspective.

> N.2 could be made optional by having a checkbox in the dom0 confirmation dialog an passing a "custom folder" parameter to the receiving agent.

This does not make sense for all rpc actions. How do you propose to distinguish between those actions for which it makes sense to show and those for which it does not?

If we really want to do this because UX, then I think a qubes.FilecopyCustomDest would be better than a checkbox, since it allows control via policy (for example: never allow custom dest dirs, except when target is "photos" or "music" VMs... or something). I'm still not convinced we want this at all though.

> 03. that no confirmation or success message is displayed in the destVM, with a link to the destination file?

Personally, I use qvm-{copy,move}-to-vm from the command line all the time, and popping up a file manager in the destvm would be rather annoying. I could be convinced that other behavior is in the best interest of the majority of users, but again it'd need to be a stronger argument than "because we can".

Note also that all the above is just my view/opinion, and I am not in charge here.

P.S.: @boring-stuff thank you very much for taking this on!

**Comment by boring-stuff** [comment]
Ok, nr.1 is settled as not wanted, then, as it needs a complete overhaul of the qrexec design first, if such a property is desired at all.

Regarding nr.2:
> sourcevm obviously must not. destvm may, but it's still dangerous. IMO the UX improvement isn't worth the danger of files ending up where they are not expected or where they have unwanted side-effects. I could be convinced otherwise, but it'd need to be a much stronger argument than "because we can". Only allowing files to land in ~/QubesIncoming/${vm}/... is a nice property from a security perspective.

Of course sourceVM should not know about the destination directory (see also [this comment](https://github.com/QubesOS/qubes-issues/issues/1793#issuecomment-196982196)), neither should dom0 (how could/why should dom0 know about the available directories in a VM?), I'm talking about the destVM which is about to unpack the file(s). 

In this case, from a usability perspective:
1. I expect a non-experienced user to transfer files for a specific purpose (i.e. storing a file from an email into an archive VM), into a specified folder structure in its home directory or attached usb disk.
2. The current workflow always sees the file being stored in the standard `QubesIncoming` folder structure, and the user having to navigate to it and copy/move it to the target directory
3. An usability target is also to reduce the confirmation dialogs

Whereas from a security perspective:
1. The `qfile-unpacker` should run non-privileged (is this the case already?), to avoid potentially overwriting system files
2. The user should be aware of the source VM at all times, to prevent a VM slipping files into another without being detected
3. Any file overwriting must be either denied (current behavior) or confirmed by the user in the destination VM (technically feasible, not implemented)

I see the current workflow satisfying the security properties (the non-privileged having to be confirmed), I have no doubts about that. What I'm proposing is the following workflow:

- User starts a transfer in sourceVM
- User confirms the transfer in dom0, specifying the desire for an alternate destination folder 
- A folder selection window pops up in the target VM asking for the destination directory

The option for an alternate directory would only be passed through the confirmation dialog  (no `allow` qrexec policy), thus satisfying security property 2., whereas the non-privileged execution would prevent overwriting system files, and file overwrite could be confirmed in the targetVM.

Am I missing something, or how likely is the usecase to be actually used in real life?

As for the file transfer checkbox:
> This does not make sense for all rpc actions. How do you propose to distinguish between those actions for which it makes sense to show and those for which it does not?

By the qrexec action name. The Qubes team could decide, as a policy, that the `qubes.*` actions are under control of the Qubes team, and custom rules should get a different prefix.
If this is true, then it would be desirable to ask some parameters to the user in dom0 (because the sourceVM cannot be trusted with the parameters), and the policy dialog would be the perfect place for this.

> If we really want to do this because UX, then I think a qubes.FilecopyCustomDest would be better than a checkbox, since it allows control via policy (for example: never allow custom dest dirs, except when target is "photos" or "music" VMs... or something). I'm still not convinced we want this at all though.

I don't agree: the user would be required to edit the policies to enable this. After that, the sourceVM could effectively pass the parameter and eventually bypass user confirmation (in case of misconfiguration, as an example). 

Regarding the confirmation message:
> Personally, I use qvm-{copy,move}-to-vm from the command line all the time, and popping up a file manager in the destvm would be rather annoying. I could be convinced that other behavior is in the best interest of the majority of users, but again it'd need to be a stronger argument than "because we can".

I agree that this is not a usecase most tech-savy users would like. I reckon the balance would be to have it happen only for the transfers which are visually confirmed through the dialog, and having a configuration file turning this off, potentially through the filecopy policy file or in a configuration file in the target VM.

Let me know how I can help at best! I've just been browsing a couple of bugs related to this...

**Comment by marmarek** [comment]
#1793 was closed for a reason. Anyway, it is offtopic in this ticket.

BTW, if you really want, you can use `mount --bind` to connect a different directory at `~/QubesIncoming/some-name`.

**Comment by marmarek** [comment]
> @marmarek I've made a first pass already, just blocking on available free time. Hopefully I'll get a chance tomorrow.

Do you mean you haven't posted in anywhere yet, or I can't find it? ;)

**Comment by jpouellet** [comment]
As in haven't posted yet, because haven't had time to look at as thoroughly as I would like. In the interest of moving things along, my notes so far are as follows:

- BaseException seems too general an exception to catch
- prefer dest whitelist to dest blacklist
- not clear that excluding self & dom0 is the right thing to do
- possible to smuggle from untrusted into gui %s?
- rpc_text: use .partition() instead of find in index -- breaks if rpc without . in name?
- really no suitable gtk oneshot timer?
- empty <placeholder/> purpose? (for client decoration buttons perhaps? idk... i'm not familiar enough with gtk, but it seems odd)
- _can_perform_action?
- keysym hard-coded?
- trailing spaces

**Comment by boring-stuff** [comment]
Thanks for the first review!

> - prefer dest whitelist to dest blacklist

That can change for the `VMListModeler`, depending on design and usage decision. Does `qrexec-policy` need/want to retrieve the list of all qubes? `VMListModeler` is meant to be re-used as well, so it does load the list by itself anyway. 

> - not clear that excluding self & dom0 is the right thing to do

I assumed sourceVM shouldn't be set as the destVM, nor dom0, but do we have usecases for this? Note that we could exclude them from the combobox but re-enable them for selection with a checkbox under the "Advanced" tab (present in the .glade, but not visible)

> - is it possible to smuggle from untrusted into gui via %s?

Good point. The variables in play are the source domain and rpc label coming from the caller qrexec-policy, and the qubes list 

> - _can_perform_action?

This internal function determines whether an action (confirm, cancel) should be allowed or not. Needed as a function, so that the test mock [can override it](https://github.com/boring-stuff/qubes-core-admin/blob/1d96f4adb1707e205d96b7f78c1fee2c226565e2/tests/rpcconfirmation.py#L67-L68) (allowing every action and avoid asking the actual focus stealing helper).

> - really no suitable gtk oneshot timer?

That's the only one I could find, but it's a minor util which could be swapped - the tests for it are also quite extensive I believe

> - empty \<placeholder/\> purpose? (for client decoration buttons perhaps? idk... i'm not familiar enough with gtk, but it seems odd)

That was generated by glade, I haven't edited the xml myself. The "advanced options" expander should also be removed if not needed.

As for the next points, see the [latest commit(s)](https://github.com/boring-stuff/qubes-core-admin/commit/1d96f4adb1707e205d96b7f78c1fee2c226565e2).
> - BaseException seems too general an exception to catch
> - rpc_text: use .partition() instead of find in index -- breaks if rpc without . in name?
> - keysym hard-coded?
> - trailing spaces

**Comment by jpouellet** [comment]
Also, the list of VMs in the dropdown should be sorted.

**Comment by jpouellet** [comment]
@boring-stuff https://github.com/jpouellet/qubes-core-admin/commit/aaa5037ad9036c713623a640a5ed329aa558ed30 adds your stuff to the qubes-core-dom0 rpm

**Comment by jpouellet** [comment]
Even though it's filtered beforehand, I don't like the fact that:
```
confirm_rpc('sys-net', 'qubes.<a href="file:///etc">foo</a>')
```
produces a clickable link which causes an appropriate handler program to get executed in dom0.

The fact that potentially untrusted (we can not predict what future gtk returns) error messages [may be smuggled into the UI the same manner](https://github.com/boring-stuff/qubes-core-admin/blob/1d96f4adb1707e205d96b7f78c1fee2c226565e2/core/rpcconfirmation.py#L74-L76) is not something I like either.

**Comment by boring-stuff** [comment]
That's true, "I just assumed..." - but then, that's how most errors are made.

[This commit](https://github.com/QubesOS/qubes-core-admin/commit/777880cb702613f4a6314434c76938bfe136d6c9) should implement proper stripping of special chars and escaping with the utility that [the docs](https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Label.html#Gtk.Label.set_markup) recommend.

I don't see a problem with `set_text `on a label, however, since the function [ignores any markup](https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Label.html#Gtk.Label.set_text) (the case for the error bar).

**Comment by boring-stuff** [comment]
I'll look into sorting later on. 
I've added you as contributors to my fork, in case you'd rather commit changes instead of proposing them in the comments... @jpouellet your comments were integrated as well.

**Comment by boring-stuff** [comment]
List sorted and whitelist of domain names can be passed [via the main API](https://github.com/boring-stuff/qubes-core-admin/blob/24f337e28c19b088afea3799add0f0d9af4b3726/core/rpcconfirmation.py#L197,L210). 

The result either `False`, if the dialog was cancelled, or a `dict` with the selected target's data.

**Comment by jpouellet** [comment]
Just double checked the sanitization code. It is done [like this](https://github.com/QubesOS/qubes-core-admin-linux/blob/4f0878ccbf8a95f8264b54d2b6f4dc433ca0793a/qrexec/qrexec-daemon.c#L627-L646) by qrexec-daemon [here](https://github.com/QubesOS/qubes-core-admin-linux/blob/4f0878ccbf8a95f8264b54d2b6f4dc433ca0793a/qrexec/qrexec-daemon.c#L669-L671) before being passed to qrexec-policy.

I'm having second thoughts about whether or not to want to do [additional filtering](https://github.com/boring-stuff/qubes-core-admin/commit/777880cb702613f4a6314434c76938bfe136d6c9#diff-ce15aa2f15b3db6535be051edca2c19dR109) here, because we also want to avoid mismatches in determinations of what should be escaped in order to avoid the possibility of a mismatch between what is displayed here and how it's interpreted elsewhere...

Disagreeing with my previous comment: perhaps it's best to just add cross-referenced comments between this code and the qrexec-daemon code noting that we rely on the filtering there, and don't attempt additional string mangling here. @boring-stuff @marmarek thoughts?

**Comment by marmarek** [comment]
> edit: To be clear: I'm proposing that it might be preferable to do GLib.markup_escape_text(), and not any further underscore substitution, but I'm not sure.

👍 

As for VM list filtering - the whitelist should be passed at API level (as it is currently). I'd make this parameter required. And not have any additional VM list filtering in GUI code.

**Comment by jpouellet** [comment]
> As for VM list filtering - the whitelist should be passed at API level (as it is currently). I'd make this parameter required. And not have any additional VM list filtering in GUI code.

+1

[Clean separation of responsibilities](http://www.catb.org/esr/writings/taoup/html/ch01s06.html#id2877777) between qrexec-policy & code responsible for user interaction is desirable. I can easily imagine someone wanting to replace the gui with something custom (for example: maybe some kind of two-factor auth hook, or some kind of "ask supervisor's machine to confirm & forward response back").

**Comment by boring-stuff** [comment]
Thanks for the references to the sanitization code! I agree that `GLib.markup_escape_text()` should stay no matter what, to avoid any problems in the future. 

However, I suggest removing the extra filtering from the GUI code only if we can guarantee the following:

- The GUI will only be called by `qrexec-policy`
- The parameter passed to `qrexec-policy` has been sanitized already ([not the case](https://github.com/QubesOS/qubes-core-admin-linux/blob/master/qrexec/qrexec-policy#L170))

I recommend creating a `sanitize_name` function in some qubes utils package that mirrors the `sanitize_name` function in the qrexec daemon (and comments referencing themselves), and be either called from the qrexec-policy or from both (ideally at least once for every module in which we handle names).

As for the additional filtering of `dom0` and the source domain, I'll remove it shortly with the next commits. Can I assume that the passed whitelist will never be empty? For that case, an error message would be more appropriate - I can display one in the window's error bar , or you can avoid calling the window at all and show an error dialog instead.

**Comment by jpouellet** [comment]
Double-sanitizing has the effect of making it unclear what exactly what component has intended responsability for sanitization. By having it in both places, we increase the chance that there will one day exist unsanitized code paths, because when making modifications we may think "no... that other place does sanitization" when it may not under all circumstances, or may be modified and/or removed later thinking "no *that* other place is where we sanitize". Accidents happen...

Perhaps a better thing than double-sanitizing would be to assert that we are already sanitized (sanitize(input) == input) in places like this where we expect to only have already-sanitized input, and error as loudly as possible if that assertion fails so that if we majorly fuck up, hopefully someone would report it and we'd realize quickly instead of going unnoticed because it's getting sanitized in (*hopefully* all) code paths further down.

**Comment by jpouellet** [comment]
Also, what makes you say the following:
> The parameter passed to qrexec-policy has been sanitized already (not the case)

?

Based on the qrexec-daemon code linked above and local testing, it appears that it is. Is there some place qrexec-policy is invoked via something other than qrexec-daemon that I am not aware of?

**Comment by jpouellet** [comment]
Also, one could imagine wanting to prompt with more human-meaningful names for the operations in the future, letting qrexec-policy keep track of what RPC operation was actually requested, and displaying some corresponding non-qubes.RPCName string.

**Comment by boring-stuff** [comment]
The key question is about which coding best practice should be followed for code handling sensitive data, e.g. when to call sanitization.
Must a string be sanitized once, even if jumping through different modules and being passed through different programs, or always for every usage of the string, or should the assertion you recommended be in place instead? This should be one of the golden rules of developing qubes, and it should be enforced.

@jpouellet I see qrexec-policy as a python program with a `__main__`, and I have no idea how it is called, but it could be called today independently with unsafe strings, and somebody in the future might forget about that and call it without sanitizing them first. 
But maybe that's exactly the key: every piece of code that may be called independently as an executable (every code with a main) must do sanitization of the inputs, no matter what we assume about the caller.
The same goes, of course, for code that handles external inputs such as files or network connections. As for code that assumes sanitized input, the parameters could have the `sanitized_` prefix to make clear to the caller that sanitized input is expected?

Applying this rule to the code in question, only the qrexec-policy should implement sanitization, as the gui has no main anyway. How does that sound?

**Comment by marmarek** [comment]
Arguments are sanitized just after retrieving them from "untrusted world". No program in dom0 is called with arguments not sanitized already - at least at the level of "strange characters".  

@jpouellet Yes, those two are the only places where qrexec-policy is called.

> Applying this rule to the code in question, only the qrexec-policy should implement sanitization, as the gui has no main anyway. How does that sound?

The risk of desynchronized sanitizers (described by @jpouellet ) is real. If anything at all, there should be only `assert(sanitize(input) == input)`.

**Comment by boring-stuff** [comment]
I see. How's [this](https://github.com/boring-stuff/qubes-core-admin/commit/ffb1357571a9c484ff85cf9faae9ba8c7a25120e#diff-ce15aa2f15b3db6535be051edca2c19dR131) then for an assertion? It could be as easily called by `qrexec-policy`, for that matter.

**Comment by jpouellet** [comment]
re service-oriented qrexec model violations:

https://github.com/QubesOS/qubes-builder-github/blob/3a38923a665c8f495d4572447332d74dba571c25/github-webhooks/trigger-build#L53-L54 is a real-world case where it makes sense to target specific destvm.

Move to service-oriented model would require multiple services (qubesbuilder.TriggerBuildDebian, qubesbuilder.TriggerBuildFedora, qubesbuilder.TriggerBuildRelease, ...).

The ideal solution would actually be to have a single dom0 service which spawns off new `$dispvm:qubes-builder-dvm-template`s per invocation, but we don't have multiple dispvm template support in a release yet.

**Comment by marmarek** [comment]
Just pushed this integrated into Qubes 4.0 code: https://github.com/marmarek/qubes-core-admin/tree/core3-policy

I've made few adjustments - mostly simplified whitelist/blacklist handling. And of course adjusted to core3 API. Some minor thins still to do:
- [ ] improve tests structure (use `setUp` instead of `__init__`, don't let tests inherit from tested classes, instead store object instance in an attribute)
- [ ] implement missing `mgmtinternal.*` methods (starting DispVM / AppVM) - this is waiting for @woju implementing `mgmt.vm.Start` method
- [ ] make sure the GUI will work even if qrexec-policy is started from root/system service:
  - for now: make sure `DISPLAY` is set, and really have access to X server
  - later: launch GUI part as separate process, inside user session (through d-bus?) - this will be required for GUI domain, where mere setting `DISPLAY` is not enough
- [ ] pylint reported bugs

@boring-stuff great work, thank you!

Warning: this branch may still be a subject to rebase, _not_ ready to be merged into `core3-devel`

**Comment by boring-stuff** [comment]
Great, thanks @jpouellet and @marmarek for the coaching, I'm glad I could help!

**Comment by marmarek** [comment]
> later: launch GUI part as separate process, inside user session (through d-bus?) - this will be required for GUI domain, where mere setting DISPLAY is not enough

Chosen this option right now - it was just few lines of code.
